src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Controls/ProgressBar.cs

```diff
@@ -274,7 +274,11 @@ internal override void ChangeVisualState(bool useTransitions)
         {
             // We change the visual state to determinate when not IsVisible for performance reasons.
             // By default, the animation for the Indeterminate state will continue even when the progress bar is not visible.
-            if (!IsIndeterminate || !IsVisible)
+
+            // However, we can only allow ourselves to do that if the ProgressBar is not being cached, since when the UIElement is defined within
+            // resources, IsVisible property will always evaluate to false as the UIElement is not a part of the visual tree at that point. (#8960)
+
+            if (!IsIndeterminate || (!IsVisible && CacheMode is null))
```

Comment:
Wouldn't it be better to check if the `ProgressBar` is part of the visual tree instead of checking `CacheMode` ?

Something like this:
```suggestion
            if (!IsIndeterminate || (!IsVisible && VisualTreeHelper.GetParent(this) is not null))
```
<br/>
I haven't tested it but to me it's more precise because the bug does not seem specific to `CacheMode`.

Reply:
@ThomasGoulet73 Maybe I scoped too narrowly indeed, I'll check this but imho should work indeed.